user, show, show details, reservation 기능 추가

taesukang-dev opened the pull request:
## 📌Linked Issues
- #19 
- #20 
- #26 
- #27 
- #30 

## ✏Change Details
- security 설정 및 로그인 추가
- show 관련 기능 추가
- show detail 관련 기능 추가
- seat 관련 기능 추가
- reservation 관련 기능 추가

## 💬Comment
- 작업하며 각각의 피쳐가 필요해 rebase 를 하다보니 많은 양을 한번에 올리게 되었습니다.
- 앞으로 남은 기능은 사용자와 판매자의 현황 관련된 기능입니다.
- 제외하고는 거의 모든 기능이 완료되었다고 말씀드릴 수 있을 것 같습니다.
- mvp 가 끝나면 동시성과 인프라, db partition, replication 등의 구체적인 설계가 필요합니다.

## 📑References
- https://docs.spring.io/spring-security/reference/servlet/authentication/architecture.html

## ✅Check List
- [x] 추가한 기능에 대한 테스트는 모두 완료하셨나요?
- [x] 코드 정렬(Ctrl + Alt + L), 불필요한 코드나 오타는 없는지 확인하셨나요?

sungyujeon committed:
chore: 누락된 seatRepository 추가

sungyujeon committed:
[#20] feat: security 설정 및 로그인 추가

taesukang-dev committed:
[#19]: BaseEntity 타입 수정

data 타입을 수정하였습니다.

taesukang-dev committed:
[#19] fix: errorcode 의 status code 를 수정하였습니다.

taesukang-dev committed:
[#19] feat: user service 의 user id 로 검색하는 로직 추가

show 기능을 추가하며 필요에 의해 추가하였습니다.

taesukang-dev committed:
[#19] fix: ddl data type 수정

auditing 을 수정하며 column type 수정

taesukang-dev committed:
[#19] feat: show repository 기능 추가

show repository 에 삭제 되지 않은 게시글을 검색하는 로직을 추가하였습니다.

taesukang-dev committed:
[#19} feat: show service 추가

기획대로 show service 를 추가하였습니다.
차후 show details 로 확장할 예정입니다.

taesukang-dev committed:
[#19] feat: show controller 구현

show controller 와 필요한 dto 들을 정의하였습니다.

taesukang-dev committed:
[#19] style: log 삭제

log 를 삭제하였습니다.

taesukang-dev committed:
[#19] style: log 삭제

log 를 삭제하였습니다.

sungyujeon committed:
[#26] feat: userDetail CRUD 구현

sungyujeon committed:
[#27] feat: seat 생성, 수정, 예약, 취소, 조회 기능 추가

sungyujeon committed:
chore: merge conflict

taesukang-dev committed:
[#19]: fix: 컨트롤러 인자 변경

get 요청에 들어오는 requestbody -> requestpram 으로 변경

taesukang-dev committed:
[#19] feat: show 도메인에 해당하는 show detail 추가하는 로직

show 도메인과 show detail 의 연관관계를 정의하는 로직 추가

taesukang-dev committed:
Merge remote-tracking branch 'origin/feature/show' into feature/show

taesukang-dev committed:
[#30] refactor: 예약 취소시 exception 발생하게끔 수정

기존 boolean 값을 리턴하던 코드에서 exception 이 발생하게끔 수정하였습니다.

taesukang-dev committed:
[#30] feat: reservation service 추가

예약을 생성하거나 취소하는 서비스를 추가하였습니다.

taesukang-dev committed:
[#30] reservation controller 추가

reservation controller 와 각 목적에 맞는 dto 를 추가하였습니다.

taesukang-dev committed:
[#27] feat: seat controller 추가

seat controller 를 추가하고 seat service 에서 transactional 을 추가하였습니다.

f-lab-joseph commented on src/main/java/com/imhero/config/BaseEntity.java:
```diff
@@ -12,8 +12,8 @@
 
     @CreatedBy
-    private Long createdBy;
+    private String createdBy;
```
localdatetime 등의 타입을 사용할 계획은 없나요 ?

sungyujeon replied:
BaseTimeEntity에서 createdAt, modifiedAt 시간은 LocalDateTime 타입으로 설정되어 있습니다.

f-lab-joseph commented on src/main/java/com/imhero/reservation/domain/ReservationCancelRequest.java:
```diff
@@ -0,0 +1,16 @@
+@NoArgsConstructor
+public class ReservationCancelRequest {
+    private Long seatId;
+    List<Long> ids = new ArrayList<>();
```
Set ?

sungyujeon replied:
https://github.com/f-lab-edu/IMHERO/pull/31/commits/7c0007d9c3dc3fabe69ace4c507303fc43d1e6c9 Set 자료구조로 변경 후 관련 로직 수정하였습니다.

f-lab-joseph commented on src/main/java/com/imhero/reservation/service/ReservationService.java:
```diff
@@ -0,0 +1,59 @@
+        User user = getUser(userName);
+        Seat seat = seatService.getSeatByIdOrElseThrow(reservationCancelRequest.getSeatId());
+        List<Reservation> reservations = reservationRepository.findAllById(reservationCancelRequest.getIds());
+        log.info("the reservation {}", reservations);
```
MDCFilter 는 설정돼있나요 ?

sungyujeon replied:
#33 이슈 발행

f-lab-joseph commented on src/main/java/com/imhero/reservation/service/ReservationService.java:
```diff
@@ -0,0 +1,59 @@
+        if (!reservations.get(0).getUser().getEmail().equals(user.getEmail())) {
+            throw new ImheroApplicationException(ErrorCode.UNAUTHORIZED_BEHAVIOR);
+        }
+        reservations.forEach(Reservation::cancel);
```
매번 업데이트 쿼리가 나가는건가요 ?

sungyujeon replied:
https://github.com/f-lab-edu/IMHERO/pull/31/commits/ee4fdf472fe96b75e4373cbc932fbe472a9c5a57 https://github.com/f-lab-edu/IMHERO/pull/31/commits/7c0007d9c3dc3fabe69ace4c507303fc43d1e6c9 `IN` query로 개선하였습니다.

f-lab-joseph commented on src/main/java/com/imhero/show/dto/request/ShowRequest.java:
```diff
@@ -0,0 +1,24 @@
+    private LocalDateTime showFromDate;
+    @JsonProperty("show_to_date")
+    private LocalDateTime showToDate;
+    // TODO: 구현할 것
```
todo ?

f-lab-joseph reviewed (commented):
PR 이 너무커서 한번에 보기 불가능하네요 .. ㅎ

sungyujeon committed:
chore: reservation test 실패 케이스 수정

sungyujeon committed:
[#30, #31] refactor: 좌석 예매 취소 쿼리 개선

- 기존: 영속성 컨텍스트 변경 감지를 통해 객체별 update query flush
- 변경: JPQL update ... in ... 쿼리로 변경

sungyujeon committed:
chore: ReservationCancelRequest dto 패키지로 이동

sungyujeon committed:
[#30, #31] refactor: 예매 취소 로직 변경

- reservation DEL_YN = 'Y' 업데이트 쿼리 조건 'N' 체크 및 update count return 추가
- Reservation entity cancel() 메서드 삭제
- ReservationCancelRequest type 변경(List to Set) 및 관련 로직 수정
